catboost/libs/metrics/metric.cpp

```diff
@@ -614,6 +614,115 @@ void TRMSEMetric::GetBestValue(EMetricBestValue* valueType, float*) const {
     *valueType = EMetricBestValue::Min;
 }
 
+/* Cox partial loss */
+
+namespace {
+    struct TCoxMetric final: public TNonAdditiveMetric {
+        explicit TCoxMetric(const TLossParams& params)
+            : TNonAdditiveMetric(ELossFunction::Cox, params)
+        {}
+
+        static TVector<THolder<IMetric>> Create(const TMetricConfig& config);
+
+        TMetricHolder Eval(
+            const TVector<TVector<double>>& approx,
+            TConstArrayRef<float> target,
+            TConstArrayRef<float> weight,
+            TConstArrayRef<TQueryInfo> queriesInfo,
+            int begin,
+            int end,
+            NPar::ILocalExecutor& executor) const override {
+                return Eval(To2DConstArrayRef<double>(approx), /*approxDelta*/{}, /*isExpApprox*/false, target, weight, queriesInfo, begin, end, executor);
+        }
+        TMetricHolder Eval(
+            const TConstArrayRef<TConstArrayRef<double>> approx,
+            const TConstArrayRef<TConstArrayRef<double>> approxDelta,
+            bool isExpApprox,
+            TConstArrayRef<float> target,
+            TConstArrayRef<float> weight,
+            TConstArrayRef<TQueryInfo> queriesInfo,
+            int begin,
+            int end,
+            NPar::ILocalExecutor& executor) const override;
+
+        double GetFinalError(const TMetricHolder& error) const override;
+        void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
+    };
+}
+
+TVector<THolder<IMetric>> TCoxMetric::Create(const TMetricConfig& config) {
+    return AsVector(MakeHolder<TCoxMetric>(config.Params));
+}
+
+TMetricHolder TCoxMetric::Eval(
+            const TConstArrayRef<TConstArrayRef<double>> approx,
+            const TConstArrayRef<TConstArrayRef<double>> /*approxDelta*/,
+            bool isExpApprox,
+            TConstArrayRef<float> target,
+            TConstArrayRef<float> /*weight*/,
+            TConstArrayRef<TQueryInfo> /*queriesInfo*/,
+            int /*begin*/,
+            int /*end*/,
+            NPar::ILocalExecutor& /*executor*/) const {
+    Y_ASSERT(!isExpApprox);
+
+    TMetricHolder error(2);
+    error.Stats[1] = 1;
+
+    TVector<size_t> label_order(target.ysize());
+    std::iota(label_order.begin(), label_order.end(), 0);
+    std::sort(label_order.begin(), label_order.end(), [&]
+        (size_t lhs, size_t rhs){
+            float l_label = target[lhs] >= 0 ? target[lhs] : -target[lhs];
+            float r_label = target[rhs] >= 0 ? target[rhs] : -target[rhs];
+            return l_label < r_label;
+        }
+    );
+
+    const yssize_t ndata = target.ysize();
+    double exp_p_sum = 0;
+    for (yssize_t i = 0; i < ndata; ++i) {
+        exp_p_sum += std::exp(approx[0][i]);
```

Comment:
Better then compute std::exp 2 times  use FastExpInPlace and save result  (an example catboost/libs/model/eval_processing.h:30 )

Reply:
FastExpInPlace doesn't meet CalcDers* qualifiers for approxes and approxDeltas'